src/CADL.Extension/Emitter.Csharp/src/lib/model.ts

```diff
@@ -407,27 +349,121 @@ export function getInputType(
         }
     }
 
-    function getInputTypeForArray(
-        program: Program,
-        elementType: Type
-    ): InputListType {
+    function getInputTypeForArray(elementType: Type): InputListType {
         return {
             Name: "Array",
             ElementType: getInputType(program, elementType, models, enums),
             IsNullable: false
         } as InputListType;
     }
 
-    function getInputTypeForMap(
-        program: Program,
-        key: Type,
-        value: Type
-    ): InputType {
+    function getInputTypeForMap(key: Type, value: Type): InputDictionaryType {
         return {
             Name: "Dictionary",
             KeyType: getInputType(program, key, models, enums),
             ValueType: getInputType(program, value, models, enums),
             IsNullable: false
         } as InputDictionaryType;
+    }   
+
+    function getInputModelForModel(m: ModelType): InputModelType {
+        m = getEffectiveSchemaType(program, m) as ModelType;
+        const name = getFriendlyName(program, m) ?? m.name;
+        let model = models.get(name);
+        if (!model) {
+            const baseModel = getInputModelBaseType(m.baseModel);
+            const properties: InputModelProperty[] = [];
+
+            model = {
+                Name: name,
+                Namespace: m.namespace?.name,
+                Description: getDoc(program, m),
+                IsNullable: false,
+                DiscriminatorPropertyName: getDiscriminator(program, m)?.propertyName,
+                DiscriminatorValue: getDiscriminatorValue(m, baseModel),
+                BaseModel: baseModel,
+                Properties: properties // Properties should be the last assigned to model
+            } as InputModelType;
+
+            models.set(name, model);
+
+            // Resolve properties after model is added to the map to resolve possible circular dependencies
+            addModelProperties(m.properties, properties, baseModel?.DiscriminatorPropertyName);
+
+            // Temporary part. Derived types may not be referenced directly by any operation
+            // We should be able to remove it when https://github.com/Azure/cadl-azure/issues/1733 is closed 
+            if (model.DiscriminatorPropertyName && m.derivedModels) {
```

Comment:
shall we also emit derived models as fields in the base model? in case, in the late codegen, the operation will set base model as input and we can easy to get its derived models to generate the sdk.

Reply:
No, derived models be restored in autorest.csharp without any additional input (because of possible circular references, it is actually easier).